### ISSUE TITLE
[experimental-utils] Is there plan for supporting other Node type at RuleCreator?

### ISSUE THREAD

**Issue by tyankatsu0105** [typescript-eslint/typescript-eslint#2045]
**Repro**

Hi!
I'm [eslint-plugin-gridsome](https://github.com/gridsome/eslint-plugin-gridsome)'s maintainer.
These days, I'm working on refactoring all rules with TypeScript.
But RuleCreator is not supporting other parser's Node type.

[vue-eslint-parser](https://github.com/mysticatea/vue-eslint-parser) is export [Node type](https://github.com/mysticatea/vue-eslint-parser/blob/master/src/ast/nodes.ts#L25)
I'd like to use this type at RuleCreator.

**Expected Result**

```ts
import { ESLintUtils } from "@typescript-eslint/experimental-utils";

import { AST } from "vue-eslint-parser";
export const createRule = ESLintUtils.RuleCreator<AST.Node>(
  name =>
    `https://github.com/gridsome/eslint-plugin-gridsome/blob/master/docs/rules/${name}.md`
);
```

**Actual Result**
experimental-utils focus on `TSESTree`. For example [this](https://github.com/typescript-eslint/typescript-eslint/blob/242328fa749ee4c72af93433a9bef95f329ac62f/packages/experimental-utils/src/ts-eslint/Rule.ts#L151).

**Additional Info**
If this request is difficult, I'd like to know some idea about using vue-eslint-parser and experimental-utils.

**Versions**

| package                                 | version  |
| --------------------------------------- | -------- |
| `@typescript-eslint/experimental-utils` | `2.33.0` |
| `TypeScript`                            | `3.9.2`  |
| `node`                                  | `13.2.0` |
| `npm`                                   | `6.13.1` |

**Comment by bradzacher** [comment]
There is no plans, no.

The unfortunate reality is that it's not really feasibly to do so.
Essentially all of the types that exist in this project are built around this union type here:
https://github.com/typescript-eslint/typescript-eslint/blob/a1816c941a5aaad55b7d304737db6da8ce67f523/packages/typescript-estree/src/ts-estree/ts-estree.ts#L140-L299

In order to support new node types in this union, we would have to make _every single type_ in the package generic, and properly thread that type down around.

It's no small task to set that up, and it creates a huge pain point for you as you would have to ensure that every single time you use any part of our tooling, you pass in your custom `Node` type.

It also isn't really a solvable problem. I've included an example for this in a fold below.

Ultimately there is a reason that both we and `vue-eslint-parser` defined our own types from the ground up instead of attempting to rework the `@types/eslint` and `@types/estree` definitions to be generic and build off of them:

It's just not what TS was designed to do. You end up wasting so much time attempting to fight the framework to make it work, and if you do make it work, any solution you arrive at is usually very brittle and provides horrible DevX.

<details>
<summary>Example</summary>

```ts
type Node = 
  | MyNode;
//| ...

interface MyNode {
  child: Node;
}
```

Now making it generic in the `Node` type:

```ts
type Node = 
  | MyNode;

interface MyNode<TNode extends Node = Node> {
  child: TNode;
}
```

Now let's add your new node type:

```ts
type CustomNode = 
  | Node
  | MyCustomNode;

interface MyCustomNode {
  child: CustomNode;
}
```

But there's a problem - you've used the `Node` type in your union, but all of the types in `Node` are all using the default generic value (which is `Node`). This means none of them reference your new `MyCustomNode`.

So let's make `Node` generic to let you pass your custom type around.

```ts
type Node<TNode = Node> =
  | MyNode<TNode>;
//| ...

// ...

//////

type CustomNode = 
  | Node<CustomNode>
  | MyCustomNode;

// ...
```

This looks neat, but you immediately find out that this doesn't work, and we receive the following TS error: `Type parameter 'TNode' has a circular default.`.

So then we can try to eliminate the default parameter, but you find you can't actually fix this.

If you remove the default, you have to provide the value for `TNode` everywhere.
So you codemod `Node` to be `Node<Node>`. But that doesn't work, because `Node` requires 1 generic type parameter. So then you need a solution for this which allows you to pass a non-generic union type in. I think the only solution for this is to create a non-generic union which is a copy-paste of the union (a major duplication of code).

</details>

**Comment by tyankatsu0105** [comment]
Thanks for your reply and explains.
I will try to create something simler RuleCreator.
close
